br/pkg/checkpoint/manager.go

```diff
@@ -0,0 +1,414 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package checkpoint
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/pingcap/errors"
+	backuppb "github.com/pingcap/kvproto/pkg/brpb"
+	"github.com/pingcap/tidb/br/pkg/glue"
+	"github.com/pingcap/tidb/br/pkg/storage"
+	"github.com/pingcap/tidb/pkg/domain"
+	"github.com/pingcap/tidb/pkg/parser/ast"
+)
+
+type SnapshotMetaManagerT = MetaManager[
+	RestoreKeyType, RestoreValueType, RestoreValueType, CheckpointMetadataForSnapshotRestore,
+]
+type LogMetaManagerT = MetaManager[
+	LogRestoreKeyType, LogRestoreValueType, LogRestoreValueMarshaled, CheckpointMetadataForLogRestore,
+]
+
+type tickDurationConfig struct {
+	tickDurationForFlush    time.Duration
+	tickDurationForChecksum time.Duration
+	retryDuration           time.Duration
+}
+
+func DefaultTickDurationConfig() tickDurationConfig {
+	return tickDurationConfig{
+		tickDurationForFlush:    defaultTickDurationForFlush,
+		tickDurationForChecksum: defaultTickDurationForChecksum,
+		retryDuration:           defaultRetryDuration,
+	}
+}
+
+type MetaManager[K KeyType, SV, LV ValueType, M any] interface {
+	RootPath() string
+
+	LoadCheckpointData(context.Context, func(K, LV)) (time.Duration, error)
+	LoadCheckpointChecksum(context.Context) (map[int64]*ChecksumItem, time.Duration, error)
+	LoadCheckpointMetadata(context.Context) (*M, error)
+	SaveCheckpointMetadata(context.Context, *M) error
+	ExistsCheckpointMetadata(context.Context) (bool, error)
+	RemoveCheckpointData(context.Context) error
+
+	// only for log restore
+	LoadCheckpointProgress(context.Context) (*CheckpointProgress, error)
+	SaveCheckpointProgress(context.Context, *CheckpointProgress) error
+	ExistsCheckpointProgress(context.Context) (bool, error)
+
+	LoadCheckpointIngestIndexRepairSQLs(context.Context) (*CheckpointIngestIndexRepairSQLs, error)
+	SaveCheckpointIngestIndexRepairSQLs(context.Context, *CheckpointIngestIndexRepairSQLs) error
+	ExistsCheckpointIngestIndexRepairSQLs(context.Context) (bool, error)
+
+	// start checkpoint runner
+	StartCheckpointRunner(
+		context.Context, tickDurationConfig, func(*RangeGroup[K, SV]) ([]byte, error),
+	) (*CheckpointRunner[K, SV], error)
+
+	// close session
+	Close()
+}
+
+type TableMetaManager[K KeyType, SV, LV ValueType, M any] struct {
+	se       glue.Session
+	runnerSe glue.Session
+	dom      *domain.Domain
+	dbName   string
+}
+
+func NewLogTableMetaManager(
+	g glue.Glue,
+	dom *domain.Domain,
+	dbName string,
+) (LogMetaManagerT, error) {
+	se, err := g.CreateSession(dom.Store())
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	runnerSe, err := g.CreateSession(dom.Store())
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return &TableMetaManager[
+		LogRestoreKeyType, LogRestoreValueType, LogRestoreValueMarshaled, CheckpointMetadataForLogRestore,
+	]{
```

Comment:
perhaps also ditto for other positions.

```suggestion
	return &LogMetaManagerT{
```